### ISSUE TITLE
Splitting arcs in a graph causes corruption.

### ISSUE THREAD

**Issue by olson-sean-k** [olson-sean-k/plexus#34]
Triangulating a graph constructed by subdividing a simple face appears to cause corruption. I discovered this while experimenting with the `subdivide` example, which demonstrates this function:

```rust
pub fn circumscribe<G>(face: FaceView<&mut MeshGraph<G>, G>) -> FaceView<&mut MeshGraph<G>, G>
where
    G: EdgeMidpoint<Midpoint = VertexPosition<G>> + GraphGeometry,
    G::Vertex: AsPosition,
{
    // Split each edge, stashing the vertex key and moving to the next arc.
    let arity = face.arity();
    let mut arc = face.into_arc();
    let mut splits = SmallVec::<[_; 4]>::with_capacity(arity);
    for _ in 0..arity {
        let vertex = arc.split_at_midpoint();
        splits.push(vertex.key());
        arc = vertex.into_outgoing_arc().into_next_arc();
    }
    // Split faces along the vertices from each arc split.
    let mut face = arc.into_face().unwrap();
    for (a, b) in splits.into_iter().perimeter() {
        face = face.split(ByKey(a), ByKey(b)).unwrap().into_face().unwrap();
    }
    // Return the central face of the subdivision.
    face
}
```

Using this function, the following code causes a panic when calling `triangulate`:

```rust
let mut graph = MeshGraph::<Point2<f64>>::from_raw_buffers(
    vec![NGon([0usize, 1, 2, 3])],
    vec![(-1.0 -1.0), (1.0, -1.0), (1.0, 1.0), (-1.0, 1.0)],
)
.unwrap();
let key = graph.faces().nth(0).unwrap().key();
// Subdivide the face twice.
circumscribe(circumscribe(graph.face_mut(key).unwrap()));
graph.triangulate(); // PANIC!
```

So far, it seems that the subdivisions work as intended. The number of vertices and faces is correct and so is the arity of the faces. However, triangulation suddenly fails when it encounters a face with arity 11. That's very strange:

1. Triangulation is implemented using splits, which should be independent of other faces; the arity of faces elsewhere in the graph should **not** be affected.
2. After the subdivisions, there are only 12 vertices in the graph, so a face with arity 11 would include all but one of them.

I suspect that something is wrong in the mutation API (specifically `FaceSplitSnapshot` and `split_with_snapshot`), but I'm not sure. For example, maybe the snapshot is failing to identify the perimeters for the two newly inserted faces for the split.

**Comment by olson-sean-k** [comment]
There are duplicate keys in the perimeters constructed by `FaceSplitSnapshot::snapshot`. This causes [this error](https://github.com/olson-sean-k/plexus/blob/01066d7d4965dae17a9a91c4d07cdb7325c9bf52/plexus/src/graph/mutation/face.rs#L256) to be returned.

**Comment by olson-sean-k** [comment]
It looks like a face insertion may be the problem. During a split, two face insertions that each use a perimeter with three vertices are performed. After this operation, an unrelated face is corrupted, and reports having 11 interior arcs. The arcs form a consistent ring, but are clearly nonsense. This could be a deeper problem with arc connectivity...

**Comment by olson-sean-k** [comment]
It seems that some vertices never have a leading arc despite being a part of a ring. Looking at the connectivity set of vertices during face insertions, there are vertices with no connectivity that are part of the connectivity of other vertices! Sure enough, trying to access the leading arc of all vertices in the graph also leads to a panic:

```rust
for vertex in graph.vertices() {
    let arc = vertex.into_outgoing_arc(); // PANIC!
    let _ = arc.key();
}
```

**Comment by olson-sean-k** [comment]
Implementing #17 likely would have caught this. Test coverage and exercising of the mutation API is not very good, but changing the [`commit` implementation](https://github.com/olson-sean-k/plexus/blob/01066d7d4965dae17a9a91c4d07cdb7325c9bf52/plexus/src/graph/mutation/vertex.rs#L59) for `VertexMutation` already causes test failures for composite edge splits:

```rust
fn commit(self) -> Result<Self::Mutant, Self::Error> {
    let VertexMutation {
        storage: vertices, ..
    } = self;
    // Fail if any vertex payloads have no leading arc. FAILS.
    let _ = vertices
        .iter()
        .map(|(_, vertex)| vertex.arc.ok_or_else(|| GraphError::TopologyMalformed))
        .collect::<Result<Vec<_>, _>>()?;
    Ok(Core::empty().bind(vertices))
}
```

**Comment by olson-sean-k** [comment]
I believe [this is the problematic line of code](https://github.com/olson-sean-k/plexus/blob/01066d7d4965dae17a9a91c4d07cdb7325c9bf52/plexus/src/graph/mutation/edge.rs#L550). It is likely more correct to disconnect vertices from their leading arcs, but the code in `split_with_cache` will leave the vertex _B_ without a leading arc when splitting from the arc _AB_.

**Comment by olson-sean-k** [comment]
This is fixed by 6bfe175 on the [`mutate` branch](https://github.com/olson-sean-k/plexus/tree/mutate). I plan to do more work on that branch to improve the `mutation` module.

**Comment by olson-sean-k** [comment]
I've cherry-picked 043abfe from the `mutate` branch into `master`. That change provides the basic fix. It was lingering on the `mutate` branch for a while, but since this is an easy bug to encounter, I wanted to pull it out of `mutate` so that it doesn't get held up by other changes.
